File: java/de.raphaelmuesseler.financer.client.javafx/src/main/java/de/raphaelmuesseler/financer/client/javafx/components/DoubleField.java
package de.raphaelmuesseler.financer.client.javafx.components;

import com.jfoenix.controls.JFXTextField;
import javafx.scene.control.TextFormatter;
import javafx.util.StringConverter;

import java.util.function.UnaryOperator;
import java.util.regex.Pattern;

public class DoubleField extends JFXTextField {
    public DoubleField() {
        super();

        Pattern validEditingState = Pattern.compile("-?(([1-9][0-9]*)|0)?([[.],][0-9]*)?");

        UnaryOperator<TextFormatter.Change> filter = c -> {
            String text = c.getControlNewText();
            if (validEditingState.matcher(text).matches()) {
                return c;
            } else {
                return null;
            }
        };

        StringConverter<Double> converter = new StringConverter<>() {

            @Override
            public Double fromString(String s) {
                if (s.isEmpty() || "-".equals(s) || ".".equals(s) || "-.".equals(s)) {
                    return 0.0;
                } else {
                    return Double.valueOf(s.replace(",", "."));
                }
            }

            @Override
            public String toString(Double d) {
                return d.toString();
            }
        };

        this.setTextFormatter(new TextFormatter<>(converter, 0.0, filter));
    }

    public void setValue(double value) {
        this.setText(Double.toString(value));
    }

    public double getValue() {
        return Double.valueOf(this.getText());
    }
}

File: java/org.financer.client.javafx/src/main/kotlin/org/financer/client/javafx/components/DoubleField.kt
package org.financer.client.javafx.components

import com.jfoenix.controls.JFXTextField
import javafx.scene.control.TextFormatter
import javafx.util.StringConverter
import java.util.function.UnaryOperator
import java.util.regex.Pattern

class DoubleField : JFXTextField() {

    var value: Double
        get() = java.lang.Double.valueOf(text)
        set(value) {
            text = java.lang.Double.toString(value)
        }

    init {
        val validEditingState = Pattern.compile("-?(([1-9][0-9]*)|0)?([[.],][0-9]*)?")
        val filter = label@ UnaryOperator<TextFormatter.Change> { c: TextFormatter.Change ->
            val text = c.controlNewText
            if (validEditingState.matcher(text).matches()) {
                return@label c
            } else {
                return@label null
            }
        }
        val converter: StringConverter<Double> = object : StringConverter<Double>() {
            override fun fromString(s: String): Double {
                return if (s.isEmpty() || "-" == s || "." == s || "-." == s) {
                    0.0
                } else {
                    java.lang.Double.valueOf(s.replace(",", "."))
                }
            }

            override fun toString(d: Double): String {
                return d.toString()
            }
        }
        textFormatter = TextFormatter(converter, 0.0, filter)
    }
}